cep-draft.md

```diff
@@ -0,0 +1,125 @@
+<table>
+    <tr><td> Title </td><td> Optional Dependencies</td>
+    <tr><td> Status </td><td> Draft </td></tr>
+    <tr><td> Author(s) </td><td> Travis Hathaway &lt;travis.j.hathaway@gmail.com&gt;</td></tr>
+    <tr><td> Created </td><td> June 7, 2023</td></tr>
+    <tr><td> Updated </td><td> June 7, 2023</td></tr>
+    <tr><td> Discussion </td><td> NA </td></tr>
+    <tr><td> Implementation </td><td> NA </td></tr>
+</table>
+
+[python-extras]: https://peps.python.org/pep-0508/#extras
+[cargo-features]: https://doc.rust-lang.org/cargo/reference/features.html
+[original-cep]: https://github.com/conda-incubator/ceps/pull/9
+[conda-issue-7502]: https://github.com/conda/conda/issues/7502
+[conda-issue-11053]: https://github.com/conda/conda/issues/11053
+[mamba-issue-245]: https://github.com/mamba-org/boa/issues/245
+
+## Abstract
+
+This CEP proposes a new method for including optional dependencies when installing conda packages.
+Optional dependencies are largely inspired by [Python's "extras"][python-extras] and
+[Cargo's "features"][cargo-features]. When implemented, package builders will have a  way to specify
+dependencies which are not required but if installed would enable different features that are
+not included by default.
+
+## Motivation
+
+Optional dependencies as a feature has been requested for conda compatible package managers for quite
+some time (e.g. [conda-7502][conda-issue-7502], [conda-11053][conda-issue-11053], and [mamba-245][mamba-issue-245]).
+Adding this feature to the conda ecosystem would also give package managers greater flexibility when
+instructing users how to install and use their software. For example, a package could support multiple
+variants by using this methodology. If a package supported either `seaborn` or `matplotlib`,
+users could specify either `seaborn` or `matplotlib` as an optional dependency.
+
+
+## Specification
+
+This section will first show how this works from the end user point of view and then talks about
+how it will work for package builders. Updates to the `repodata.json` format will also be discussed.
+
+### User point of view
+
+When a user goes to install a package, they will specify optional dependencies with the following syntax:
+
+```bash
+# For a single optional dependency
+conda install package_name[optional_a]
+
+# For multiple optional dependencies
+conda install package_name[optional_a,optional_b]
+```
+
+Optional dependencies can also be specified by using the keyword argument syntax:
+
+```bash
+# For a single optional dependency
+conda install package_name[extras=optional_a]
+
+# For multiple optional dependencies
+conda install package_name[extras=optional_a,optional_b]
```

Comment:
How does the parsing here work? It looks like it would break the `*args,**kwargs` pattern proposed in https://github.com/conda/conda/issues/11053 (which you're referencing already).

I'm also wondering if we really _need_  two ways to do this? `package_name[optional_a,optional_b]` looks perfectly alright to me, and matches with what pip is doing.

While it's out of scope for this CEP, I'd really like to keep the door open to eventually extend the syntax to be able to use (exposition only, all kwargs need a proper design)
```
conda install package_name[optional_a,optional_b,cuda=True,license=no_gpl,cpu=avx512,...]
```
It seems to me that if we add the `extras=` way to specify optional dependencies, we don't gain much and risk making it much harder to eventually solve that problem.

Reply:
+1 for following the pip syntax.